compiler/hash-typecheck/src/traverse/mod.rs

```diff
@@ -2138,27 +2148,40 @@ impl<'tc> visitor::AstVisitor for TcVisitor<'tc> {
         node: hash_ast::ast::AstNodeRef<hash_ast::ast::BindingPat>,
     ) -> Result<Self::BindingPatRet, Self::Error> {
         let name = node.name.ident;
-        let term = self.builder().create_var_term(name);
+        let var_term = self.builder().create_var_term(name);
 
-        match self.scope_manager().resolve_name_in_scopes(name, term) {
-            Ok(_) => Ok(self.builder().create_pat(Pat::Const(ConstPat { term }))),
-            Err(_) => {
-                let pat = self.builder().create_binding_pat(
-                    node.name.body().ident,
-                    match node.mutability.as_ref().map(|x| *x.body()) {
-                        Some(hash_ast::ast::Mutability::Mutable) => Mutability::Mutable,
-                        Some(hash_ast::ast::Mutability::Immutable) | None => Mutability::Immutable,
-                    },
-                    match node.visibility.as_ref().map(|x| *x.body()) {
-                        Some(hash_ast::ast::Visibility::Private) | None => Visibility::Private,
-                        Some(hash_ast::ast::Visibility::Public) => Visibility::Public,
-                    },
-                );
-                self.copy_location_from_node_to_target(node, pat);
+        if let Ok(scope_member) = self.scope_manager().resolve_name_in_scopes(name, var_term) {
+            let term = scope_member.member.value_or_ty();
+            let ty = scope_member.member.ty();
+
+            // So this should only become a constant if we are a `enum` or `struct`
+            // definition. Otherwise, we shadow the variable
+            if self.oracle().term_is_struct_def(term) || self.oracle().term_is_enum_def(term) {
+                // @@Hack: we technically want to use the `term` from the scope here, but
+                // because the locations are messed up, instead we just copy the
+                // location to the created term
```

Comment:
What do you mean by "the locations are messed up"?

Reply:
I think the comment is now inaccurate, since the term location should originate from the type that's of the variable. However, it doesn't matter since it will be simplified immediately in `visit_constructor_term` which then takes the struct/enum def as the term and thus the same location. So actually, this location copy could probably be removed

Reply:
actually this is not a correct statement, the copy is necessary but the comment is off so I am going to remove it.